File: packages/apple/StellarTrailKit/Sources/StellarTrailKit/Domain/SkillModels.swift
import Foundation

struct SkillCategorySummary: Codable, Equatable, Identifiable {
    let id: String
    let slug: String?
    let title: String
    let summary: String
    let itemCount: Int
    let href: String?
}

struct SkillCategoriesResponse: Codable, Equatable {
    let items: [SkillCategorySummary]
}

struct PageInfo: Codable, Equatable {
    let limit: Int
    let offset: Int
    let nextOffset: Int?
}

struct KnotTaxonomyItem: Codable, Equatable, Identifiable {
    let id: String
    let slug: String?
    let title: String
}

struct KnotMediaAsset: Codable, Equatable, Identifiable {
    let id: String
    let mediaType: String
    let url: String
    let mimeType: String
    let width: Int?
    let height: Int?
    let sizeBytes: Int
    let attribution: String?
    let licenseNote: String?
}

struct KnotSummary: Codable, Equatable, Identifiable {
    let id: String
    let slug: String?
    let title: String
    let summary: String
    let difficulty: String?
    let categories: [KnotTaxonomyItem]
    let types: [KnotTaxonomyItem]
    let media: [KnotMediaAsset]
    let href: String?

    var mediaCount: Int { media.count }

    enum CodingKeys: String, CodingKey {
        case id, slug, title, summary, difficulty, categories, types, media, href
        case mediaCount
    }

    init(id: String, slug: String?, title: String, summary: String, difficulty: String?, categories: [KnotTaxonomyItem], types: [KnotTaxonomyItem], media: [KnotMediaAsset], href: String?) {
        self.id = id
        self.slug = slug
        self.title = title
        self.summary = summary
        self.difficulty = difficulty
        self.categories = categories
        self.types = types
        self.media = media
        self.href = href
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        slug = try container.decodeIfPresent(String.self, forKey: .slug)
        title = try container.decode(String.self, forKey: .title)
        summary = try container.decode(String.self, forKey: .summary)
        difficulty = try container.decodeIfPresent(String.self, forKey: .difficulty)
        categories = try container.decodeIfPresent([KnotTaxonomyItem].self, forKey: .categories) ?? []
        types = try container.decodeIfPresent([KnotTaxonomyItem].self, forKey: .types) ?? []
        media = try container.decodeIfPresent([KnotMediaAsset].self, forKey: .media) ?? []
        href = try container.decodeIfPresent(String.self, forKey: .href)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encodeIfPresent(slug, forKey: .slug)
        try container.encode(title, forKey: .title)
        try container.encode(summary, forKey: .summary)
        try container.encodeIfPresent(difficulty, forKey: .difficulty)
        try container.encode(categories, forKey: .categories)
        try container.encode(types, forKey: .types)
        try container.encode(media, forKey: .media)
        try container.encodeIfPresent(href, forKey: .href)
    }
}

struct KnotListResponse: Codable, Equatable {
    let locale: String?
    let items: [KnotSummary]
    let page: PageInfo

    var nextOffset: Int? { page.nextOffset }

    enum CodingKeys: String, CodingKey {
        case locale, items, page
        case nextOffset
    }

    init(locale: String? = "zh-CN", items: [KnotSummary], page: PageInfo) {
        self.locale = locale
        self.items = items
        self.page = page
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        locale = try container.decodeIfPresent(String.self, forKey: .locale)
        items = try container.decode([KnotSummary].self, forKey: .items)
        if let page = try container.decodeIfPresent(PageInfo.self, forKey: .page) {
            self.page = page
        } else {
            let nextOffset = try container.decodeIfPresent(Int.self, forKey: .nextOffset)
            self.page = PageInfo(limit: items.count, offset: 0, nextOffset: nextOffset)
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encodeIfPresent(locale, forKey: .locale)
        try container.encode(items, forKey: .items)
        try container.encode(page, forKey: .page)
    }
}

typealias ListKnotsResponse = KnotListResponse

struct KnotDetail: Codable, Equatable, Identifiable {
    let id: String
    let slug: String?
    let title: String
    let summary: String
    let difficulty: String?
    let categories: [KnotTaxonomyItem]
    let types: [KnotTaxonomyItem]
    let media: [KnotMediaAsset]
    let href: String?
    let description: String?
    let steps: [String]
    let locale: String?

    var mediaCount: Int { media.count }
}

struct ListKnotsRequest: Equatable {
    var offset: Int = 0
    var limit: Int = 20
    var category: String?
    var q: String?

    var queryItems: [URLQueryItem] {
        var items = [
            URLQueryItem(name: "offset", value: String(offset)),
            URLQueryItem(name: "limit", value: String(limit))
        ]
        if let category { items.append(URLQueryItem(name: "category", value: category)) }
        if let q = q?.nilIfBlank { items.append(URLQueryItem(name: "q", value: q)) }
        return items
    }
}

File: apps/android/src/main/java/com/rustella/stellartrail/domain/skills/SkillModels.kt
package com.rustella.stellartrail.domain.skills

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class SkillLocale(val headerValue: String) {
    @SerialName("zh-CN") ZH_CN("zh-CN"),
    @SerialName("en") EN("en"),
}

@Serializable
data class SkillCategorySummary(
    val id: String,
    val slug: String,
    val title: String,
    val summary: String,
    @SerialName("item_count") val itemCount: Int,
    val href: String,
)

@Serializable
data class SkillCategoriesResponse(
    val items: List<SkillCategorySummary>,
)

@Serializable
data class PageInfo(
    val limit: Int,
    val offset: Int,
    @SerialName("next_offset") val nextOffset: Int? = null,
)

@Serializable
data class KnotTaxonomyItem(
    val id: String,
    val slug: String,
    val title: String,
)

@Serializable
data class KnotMediaAsset(
    val id: String,
    @SerialName("media_type") val mediaType: String,
    val url: String,
    @SerialName("mime_type") val mimeType: String,
    val width: Int? = null,
    val height: Int? = null,
    val attribution: String? = null,
    @SerialName("license_note") val licenseNote: String? = null,
)

@Serializable
data class KnotSummary(
    val id: String,
    val slug: String,
    val title: String,
    val summary: String,
    val categories: List<KnotTaxonomyItem> = emptyList(),
    val types: List<KnotTaxonomyItem> = emptyList(),
    val media: List<KnotMediaAsset> = emptyList(),
    val href: String,
    val aliases: List<String> = emptyList(),
)

@Serializable
data class KnotListResponse(
    val locale: SkillLocale,
    val items: List<KnotSummary>,
    val page: PageInfo,
)

@Serializable
enum class FavoriteSkillCategory(val queryValue: String) {
    @SerialName("all") ALL("all"),
    @SerialName("knots") KNOTS("knots"),
}

@Serializable
data class FavoriteSkillFilterOption(
    val id: FavoriteSkillCategory,
    val title: String,
    val count: Int,
)

@Serializable
data class FavoriteKnotItem(
    @SerialName("skill_category") val skillCategory: String,
    @SerialName("favorited_at") val favoritedAt: String,
    val knot: KnotSummary,
)

@Serializable
data class ListFavoriteSkillsResponse(
    val locale: SkillLocale,
    val filters: List<FavoriteSkillFilterOption> = emptyList(),
    val items: List<FavoriteKnotItem>,
    val page: PageInfo,
)

@Serializable
data class FavoriteKnotStatusResponse(
    @SerialName("skill_category") val skillCategory: String,
    @SerialName("knot_id") val knotId: String,
    @SerialName("is_favorited") val isFavorited: Boolean,
    @SerialName("favorited_at") val favoritedAt: String? = null,
)

data class ListFavoriteSkillsRequest(
    val skillCategory: FavoriteSkillCategory = FavoriteSkillCategory.ALL,
    val offset: Int = 0,
    val limit: Int = 20,
)

@Serializable
data class KnotDetail(
    val id: String,
    val slug: String,
    val title: String,
    val summary: String,
    val categories: List<KnotTaxonomyItem> = emptyList(),
    val types: List<KnotTaxonomyItem> = emptyList(),
    val media: List<KnotMediaAsset> = emptyList(),
    val href: String? = null,
    val description: String? = null,
    val steps: List<String> = emptyList(),
    val locale: SkillLocale,
    val aliases: List<String> = emptyList(),
)

data class ListKnotsRequest(
    val offset: Int = 0,
    val limit: Int = 20,
    val category: String? = null,
    val query: String? = null,
)

fun resolveMediaUrl(assetsBaseUrl: String, mediaUrl: String): String {
    if (mediaUrl.startsWith("http://") || mediaUrl.startsWith("https://")) return mediaUrl
    if (
        mediaUrl.startsWith("android.resource://") ||
        mediaUrl.startsWith("content://") ||
        mediaUrl.startsWith("file://")
    ) {
        return mediaUrl
    }
    return assetsBaseUrl.trimEnd('/') + "/" + mediaUrl.trimStart('/')
}

fun List<KnotMediaAsset>.preferredThumbnailUrl(): String? =
    firstOrNull { it.mediaType == "thumbnail" }?.url
        ?: firstOrNull { it.mediaType == "preview" }?.url
        ?: firstOrNull { it.mimeType.startsWith("image/") }?.url

fun List<KnotMediaAsset>.preferredPreviewUrl(): String? =
    firstOrNull { it.mediaType == "preview" }?.url
        ?: firstOrNull { it.mediaType == "thumbnail" }?.url
        ?: firstOrNull { it.mimeType.startsWith("image/") }?.url
